src/main/java/com/google/devtools/build/lib/remote/blobstore/http/HttpBlobStore.java

```diff
@@ -0,0 +1,242 @@
+ * roundtrip for every upload and with little practical value as we would expect most uploads to be
+ * accepted.
+ *
+ * <p>The implementation currently does not support transfer encoding chunked.
+ */
+public class HttpBlobStore implements SimpleBlobStore {
+
+  private final NioEventLoopGroup eventLoop = new NioEventLoopGroup(8);
+  private final SimpleChannelPool downloadChannels;
+  private final SimpleChannelPool uploadChannels;
+  private final URI uri;
+
+  public HttpBlobStore(URI uri, int timeoutMillis, @Nullable final Credentials creds)
+      throws Exception {
+    boolean useTls = uri.getScheme().equals("https");
+    if (uri.getPort() == -1) {
+      int port = useTls ? 443 : 80;
+      uri =
+          new URI(
+              uri.getScheme(),
+              uri.getUserInfo(),
+              uri.getHost(),
+              port,
+              uri.getPath(),
+              uri.getQuery(),
+              uri.getFragment());
+    }
+    this.uri = uri;
+
+    final SslContext sslCtx;
+    if (useTls) {
+      sslCtx = SslContextBuilder.forClient().sslProvider(SslProvider.OPENSSL).build();
+    } else {
+      sslCtx = null;
+    }
+    Bootstrap clientBootstrap =
+        new Bootstrap()
+            .channel(NioSocketChannel.class)
+            .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, timeoutMillis)
+            .option(ChannelOption.SO_TIMEOUT, timeoutMillis)
+            .group(eventLoop)
+            .remoteAddress(uri.getHost(), uri.getPort());
+    downloadChannels =
+        new SimpleChannelPool(
+            clientBootstrap,
+            new ChannelPoolHandler() {
+              @Override
+              public void channelReleased(Channel ch) throws Exception {}
+
+              @Override
+              public void channelAcquired(Channel ch) throws Exception {}
+
+              @Override
+              public void channelCreated(Channel ch) throws Exception {
+                ChannelPipeline p = ch.pipeline();
+                if (sslCtx != null) {
+                  SSLEngine engine = sslCtx.newEngine(ch.alloc());
+                  engine.setUseClientMode(true);
+                  p.addFirst(new SslHandler(engine));
+                }
+                p.addLast(new HttpClientCodec());
+                p.addLast(new HttpDownloadHandler(creds));
+              }
+            });
+    uploadChannels =
+        new SimpleChannelPool(
+            clientBootstrap,
+            new ChannelPoolHandler() {
+              @Override
+              public void channelReleased(Channel ch) throws Exception {}
+
+              @Override
+              public void channelAcquired(Channel ch) throws Exception {}
+
+              @Override
+              public void channelCreated(Channel ch) throws Exception {
+                ChannelPipeline p = ch.pipeline();
+                if (sslCtx != null) {
+                  SSLEngine engine = sslCtx.newEngine(ch.alloc());
+                  engine.setUseClientMode(true);
```

Comment:
Doesn't this force client-side auth for the SSL connection? Do we always have an ssl cert on the client? Or maybe I misunderstand what this does?

Reply:
`sslCtx` will only be none null if --tls_enabled flag is specified. See HttpBlobStore:96.